### PR TITLE
Interpreter for DAG IR

### PR DESCRIPTION
**Pull request by oflatt**
This PR fixes up the interpreter to work with the DAG IR. It does so by caching results for evaluating nodes based on the Rc pointer so that they are evaluated just once.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -67,22 +67,20 @@ impl Optimizer {
     }
 
     /// Produces a tuple value from the string arguments
```
Comment (ajpal) [inline src/lib.rs:69]: nit: update comment

File: tree_in_context/src/interpreter.rs [hunk 2]
```diff
@@ -1,6 +1,14 @@
-use std::{collections::HashMap, fmt::Display};
+//! Interpret DAG programs. Each expression is evaluated once, so side-effects only happen
+//! once. All the dependencies of an expression are evaluted first.
+//! The interpreter relies on the invariant that common subexpressions are
```
Comment (ajpal) [inline tree_in_context/src/interpreter.rs:3]: Is this invariant checked/enforced or is this comment the only place it's clearly documented?

File: tree_in_context/src/interpreter.rs [hunk 3]
```diff
@@ -83,6 +98,8 @@ pub(crate) struct VirtualMachine<'a> {
     next_addr: usize,
     /// All of memory
     mem: HashMap<usize, Value>,
+    /// Values for already evaluated expressions
+    memo: HashMap<*const Expr, Value>,
```
Comment (ajpal) [inline tree_in_context/src/interpreter.rs:102]: nit: `mem` and `memo` are quite similar, is there a better name for one?

File: tree_in_context/src/interpreter.rs [hunk 4]
```diff
@@ -263,11 +310,14 @@ impl<'a> VirtualMachine<'a> {
...
-            Expr::Alloc(e_size, _ignore_state, _ty) => {
+            Expr::Alloc(e_size, state_expr, _ty) => {
                 let size = self.interp_int_expr(e_size, arg);
+                let state_val = self.interpret_expr(state_expr, arg);
+                assert_eq!(state_val, Value::StateV);
                 let addr = self.next_addr;
                 self.next_addr += usize::try_from(size).unwrap();
-                Ptr(Pointer::new(addr, size as usize, 0))
+
+                tuplev!(Ptr(Pointer::new(addr, size as usize, 0)), Value::StateV)
```
Comment (ajpal) [inline tree_in_context/src/interpreter.rs:320]: what's the 0 for?

File: tree_in_context/src/typechecker.rs [hunk 5]
```diff
@@ -51,19 +51,25 @@ impl Expr {
     }
 }
 
+/// A map from the old, untyped expression to the new, typed expression
+pub type TypedExprCache = HashMap<*const Expr, RcExpr>;
```
Comment (ajpal) [inline tree_in_context/src/typechecker.rs:55]: what's the difference between TypeExprCache and TypeCache?

File: tree_in_context/src/interpreter.rs (hunk 4 above)
Comment (oflatt) [reply tree_in_context/src/interpreter.rs:320]: 0 is the offset for the pointer (alloc returns a new memory region with a pointer pointing to the start of it)
Added a comment

File: tree_in_context/src/typechecker.rs (hunk 5 above)
Comment (oflatt) [reply tree_in_context/src/typechecker.rs:55]: Good question, I wrote new documentation to try to clarify
```
/// Typechecking produces new, typed expressions.
/// This map is used to memoize the results of typechecking.
/// It maps the old untyped expression to the new typed expression
pub type TypedExprCache = HashMap<*const Expr, RcExpr>;

/// We also need to keep track of the type of the newly typed expression.
/// This maps the newly instrumented expression to its type.
pub type TypeCache = HashMap<*const Expr, Type>;
```

**Review by ajpal — APPROVED** [review]

**Pull request merged**